File: solutions/javascript/list-ops/3/list-ops.js
export class List {
  constructor(values = []) {
    this.values = values;
  }

  append(list) {
    return new List([...this.values, ...list.values]);
  }

  concat(list) {
    return list.foldl((concatenated, nextList) => concatenated.append(nextList), new List(this.values));
  }

  filter(fn, values = this.values) {
    return new List(this.foldl((filtered, nextItem) => fn(nextItem) ? 
      [...filtered, nextItem] : 
      [...filtered], []));
  }

  map(fn, values = this.values) {
    return new List(this.foldl((mapped, nextItem) => [...mapped, fn(nextItem)], []));
  }

  length(values = this.values) {
    const [head, ...tail] = values;
    if (!head) return 0;
    return 1 + this.length(tail)
  }

  foldl(fun, start) {
    if (this.length() === 0) return start;
    const [first, ...rest] = this.values;
    const list = new List([...rest])
    return list.foldl(fun, fun(start, first))
  }

  foldr(fun, start) {
    return this.reverse().foldl(fun, start);
  }

  reverse(values = this.values) {
    return new List(this.foldl((reversed, nextItem) => [nextItem, ...reversed], []))
  }
}

File: solutions/python/list-ops/3/list_ops.py
def append(list1, list2):
    return list1 + [] if not list2 else list1 + [list2[0]] + append([], list2[1:])


def concat(lists):
    return [] if not lists else lists[0] + concat(lists[1:])


def filter(function, list):
    if not list:
        return []
    if not function(list[0]):
        return filter(function, list[1:])
    return [list[0]] + filter(function, list[1:])


def length(list):
    return 0 if not list else 1 + length(list[1:])
    

def map(function, list):
    return [] if not list else [function(list[0])] + map(function, list[1:])


def foldl(function, list, initial):
    return initial if not list else foldl(function, list[1:], function(initial, list[0]))


def foldr(function, list, initial):
    return foldl(function, reverse(list), initial)


def reverse(list):
    return [] if not list else reverse(list[1:]) + [list[0]]
